File: app/src/main/java/com/pin/recommend/model/entity/StoryPicture.java
package com.pin.recommend.model.entity;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

import com.pin.imageutil.BitmapUtility;
import com.pin.recommend.R;

import static androidx.room.ForeignKey.CASCADE;

@Entity(foreignKeys = @ForeignKey(entity = Story.class, parentColumns = "id", childColumns = "storyId", onDelete = CASCADE, onUpdate = CASCADE))
public class StoryPicture implements Parcelable {

    @PrimaryKey(autoGenerate = true)
    public long id;

    @ColumnInfo(index = true)
    public long storyId;

    public String uri;

    public Bitmap testGetBitmap(Context context, int width, int height){
        Bitmap bitmap = BitmapUtility.decodeResource(context, R.drawable.test, width, height);
        bitmap = BitmapUtility.getRoundedCornerBitmap(bitmap, 50);
        return bitmap;
    }

    public Drawable getBackgroundDrawable(Context context, int w, int h){
        Bitmap bitmap = BitmapUtility.readPrivateImage(context, uri, w, h);
        return new BitmapDrawable(context.getResources(), bitmap);
    }

    public Bitmap getBitmap(Context context, int width, int height) {
        Bitmap bitmap = BitmapUtility.readPrivateImage(context, uri, width, height);
        return bitmap;
    }

    public boolean saveImage(Context context, Bitmap bitmap){
        if(BitmapUtility.fileExistsByPrivate(context, uri)){
            BitmapUtility.deletePrivateImage(context, uri);
        }

        String filename = BitmapUtility.generateFilename();
        String ext = ".png";
        boolean success = BitmapUtility.insertPrivateImage(context, bitmap, filename, ext);

        uri = filename + ext;
        return success;
    }

    public boolean deleteImage(Context context){
        if(uri != null){
            BitmapUtility.deletePrivateImage(context, uri);
            this.uri = null;
            return true;
        }
        return false;
    }

    public StoryPicture() {
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(this.id);
        dest.writeLong(this.storyId);
        dest.writeString(this.uri);
    }

    protected StoryPicture(Parcel in) {
        this.id = in.readLong();
        this.storyId = in.readLong();
        this.uri = in.readString();
    }

    public static final Creator<StoryPicture> CREATOR = new Creator<StoryPicture>() {
        @Override
        public StoryPicture createFromParcel(Parcel source) {
            return new StoryPicture(source);
        }

        @Override
        public StoryPicture[] newArray(int size) {
            return new StoryPicture[size];
        }
    };
}

File: app/src/main/java/com/pin/recommend/model/entity/StoryPicture.kt
package com.pin.recommend.model.entity

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey
import com.google.gson.Gson
import com.pin.imageutil.BitmapUtility
import com.pin.recommend.R
import com.pin.recommend.model.entity.Story
import java.util.*

@Entity(
    foreignKeys = [ForeignKey(
        entity = Story::class,
        parentColumns = arrayOf("id"),
        childColumns = arrayOf("storyId"),
        onDelete = ForeignKey.CASCADE,
        onUpdate = ForeignKey.CASCADE
    )]
)
class StoryPicture {
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0

    @ColumnInfo(index = true)
    var storyId: Long = 0
    var uri: String? = null

    fun testGetBitmap(context: Context?, width: Int, height: Int): Bitmap {
        var bitmap = BitmapUtility.decodeResource(context, R.drawable.test, width, height)
        bitmap = BitmapUtility.getRoundedCornerBitmap(bitmap, 50)
        return bitmap
    }

    fun getBackgroundDrawable(context: Context, w: Int, h: Int): Drawable {
        val bitmap = BitmapUtility.readPrivateImage(context, uri, w, h)
        return BitmapDrawable(context.resources, bitmap)
    }

    fun getBitmap(
        context: Context?,
        width: Int,
        height: Int
    ): Bitmap? {
        if(uri == null) return null
        return BitmapUtility.readPrivateImage(context, uri, width, height)
    }

    fun saveImage(context: Context?, bitmap: Bitmap?): Boolean {
        if (BitmapUtility.fileExistsByPrivate(context, uri)) {
            BitmapUtility.deletePrivateImage(context, uri)
        }
        val filename = BitmapUtility.generateFilename()
        val ext = ".png"
        val success = BitmapUtility.insertPrivateImage(context, bitmap, filename, ext)
        uri = filename + ext
        return success
    }

    fun deleteImage(context: Context?): Boolean {
        if (uri != null) {
            BitmapUtility.deletePrivateImage(context, uri)
            uri = null
            return true
        }
        return false
    }

    fun toDraft(context: Context): Draft {
        return Draft(getBitmap(context, 500, 500))
    }

    class Draft(
        var bitmap: Bitmap?
    ) {
        val uuid = UUID.randomUUID()
    }
}